README.md
This Project is No Longer Maintained
===============================
Don't forget everything happened in this land in the past three years. 

-------------------
[README.bak.md](README.bak.md)


CHANGELOG.md
# 更新日志

## [TheEnd] - 2023-04-11
### Removed
- Remove Everything

## [6.2.2] - 2023-02-03
### Removed
- 移除进出校申请

## [6.2.1] - 2023-02-03
### Removed
- 移除核酸结果监控

## [6.2.0] - 2022-12-15
### Removed
- 移除图片上传功能

## [6.1.1] - 2022-12-07
### Fixed
- 修复 县区 字段变为可选的问题

## [6.1.0] - 2022-10-29
### Added
- 原 **自动申请次日离校** 功能，现增强为 **自动申请次日进校和离校**

## [6.0.2] - 2022-10-17
### Fixed
- 修复 `提交信息异常`

## [6.0.1] - 2022-10-05
### Fixed
- 修复了 `F_STATE` 未生成的问题

## [6.0.0] - 2022-10-04
### Added
- 增加 *自动申请次日离校* 功能

## [5.6.11] - 2022-09-26
### Fixed
- 适配 曾赴省市-抵沪日期

## [5.6.10] - 2022-09-22
### Added
- 适配 是否曾确诊（否）

## [5.6.9] - 2022-09-18
### Fixed
- 适配 XingCode
### Deprecated
- SuiCode 将在未来废弃

## [5.6.8] - 2022-09-12
### Added
- 适配 近7日是否曾赴外省市 选项
### Changed
- 缩短了部分休眠时间以提升性能

## [5.6.7] - 2022-08-28
### Changed
- 高中低风险地区旅居史 修改为 获取上次填写的值

## [5.6.6] - 2022-08-21
### Added
- 适配高中低风险地区旅居史（默认为无）

## [5.6.5] - 2022-07-06
### Added
- 适配申请进出校字段（假期期间默认为不需要申请）

## [5.6.4] - 2022-07-06
### Added
- 适配所在街道字段
### Removed
- 移除 XingCode 的 * 号

## [5.6.3] - 2022-06-10
### Fixed
- 完善进校校区字段

## [5.6.2] - 2022-06-09
### Added
- 适配进校校区字段

## [5.6.0] - 2022-04-28
### Added
- 增加核酸检测结果监控功能，详见 [核酸检测结果监控说明](PCR_Checker.md)

## [5.5.6] - 2022-04-15
### Changed
- 获取风险地区改为自建 API

## [5.5.5] - 2022-04-14
### Fixed
- 适配图片上传机制
- 优化抢排名功能

## [5.5.4] - 2022-03-29
### Fixed
- 修复 Actions 配置文件判断错误
- 修复登录时密码错误判断

## [5.5.3] - 2022-03-28
### Fixed
- 适配 SuiShen 码时间显示变动
- 修复阅读未读消息日志输出错误

## [5.5.2] - 2022-03-28
### Added
- 优化文档说明
- 改进部分代码

## [5.5.1] - 2022-03-28
### Added
- Actions 新增模拟随机 IP 功能，暂时可规避 IP 屏蔽
- 若填报失败或模拟 IP 失效，可继续使用连接学校 VPN 再填报功能
- 优化了一些输出提示

## [5.5.0] - 2022-03-27
### Added
- 增加学校 VPN 连接失败后重新连接的功能

## [5.4.6] - 2022-03-27
### Fixed
- 修复了数个潜在问题
### Added
- 增加了服务器使用时的填报实时输出
- 完善了文档说明

## [5.4.5] - 2022-03-25
### Fixed
- 修复了输出调试信息的字段没有及时更新的问题
### Changed
- 更正本项目用途
### Added
- 强调免责声明

## [5.4.4] - 2022-03-21
### Fixed
- 适配多个字段变动

## [5.4.3] - 2022-03-20
### Fixed
- 修复调试输出存在的潜在 bug
- 优化两码上传逻辑

## [5.4.2] - 2022-03-20
### Removed
- 移除 GitHub Actions 工作流无用命令

## [5.4.1] - 2022-03-20
### Changed
- 优化`所在校区`字段获取逻辑
- 优化两码上报逻辑，避免无用上传
- 适配 SuiShen 码变动，移除旧的 XingCheng 码样式
- 优化调试输出

## [5.4.0] - 2022-03-19
### Added
- 适配新增的`所在校区`字段
### Changed
- 优化部分代码逻辑

## [5.3.1] - 2022-03-14
### Fixed
- 以无缓冲输出方式在 GitHub Actions 上运行，修复无法及时查看日志的问题
### Added
- 登录后自动关闭是否更新档案的对话框
### Changed
- 优化多用户下的休眠间隔

## [5.3.0] - 2022-03-13
### Added
- 新增推送加PushPlus消息推送接口 [https://www.pushplus.plus/](https://www.pushplus.plus/)
### Changed
- 优化 GitHub Actions 日志输出提示
### Removed
- 由于推送加（hxtrip域名下）的消息推送接口已长期处于不可用状态，因此弃用该接口

## [5.2.0] - 2022-03-10
### Added
- GitHub Actions 增加运行时提示
- 为校内 VPN 增加路由配置，减少通过 VPN 的请求，并修复 Telegram Bot 推送不能使用的问题
- GitHub Actions 输出IP信息，方便调试

## [5.1.1] - 2022-03-05
### Added
- 新增登录结果判断，避免密码重试次数过多造成的全局填报失败

## [5.1.0] - 2022-02-28
### Added
- GitHub Actions 支持通过连接学校 VPN 后再填报，防止填报失败

## [5.0.6] - 2022-02-26
### Fixed
- 修复了上传 XingCheng 码和 SuiShen 码时返回数据的解析错误
- 修复获取风险地区时可能存在的卡死问题

## [5.0.5] - 2022-02-09
### Changed
- 改进生成的 XingCheng 码和 SuiShen 码

## [5.0.4] - 2022-02-09
### Changed
由于存在 bug ，恢复原获取地址逻辑

## [5.0.3] - 2022-02-09 [YANKED]
### Fixed
- 修复因优化了地址获取逻辑导致的某些问题

## [5.0.2] - 2022-02-09
### Changed
- 更新了基于时间选择 XingCheng 码样式

## [5.0.1] - 2022-02-08 [YANKED]
### Changed
- 优化了地址获取逻辑
- 更新 XingCheng 码样式

## [5.0.0] - 2022-02-04
### Changed
- 此版本开始增加本更新日志文件，版本号设置为 5.0.0

## 2022-02-01
### Added
- 增加填报失败时打印调试信息

## 2022-01-29
### Added
- 更好的模拟 SuiShen 码， XingCheng 码

## 2022-01-27
### Added
- 增加自动补报功能
- 新增 PushDeer 消息推送渠道

## 2022-01-25
### Changed
- 实现模拟 XingCheng 码上报

## 2022-01-15
### Added
- 支持 SuiShen 码填报（空白图）
### Fixed
- `是否在校` 字段修复

## 2021-11-17
### Changed
-  XingCheng 码上传策略改为默认每次上传空白图，同时支持手动上传提示

## 2021-11-14
### Added
- 自动阅读消息，并将必读消息的标题进行推送

## 2021-11-05
### Added
- 答题功能适配

## 2021-11-04
### Changed
-  XingCheng 码上传策略改为默认上传上一次上传的图

## 2021-10-29
### Added
- 增加当需要上传 XingCheng 码时，可提示手动上传 XingCheng 码
### Deprecated
-  SuiShen 码现无需提交，未来可能移除

## 2021-10-25
### Added
- 增加 Telegram Bot 消息推送

## 2021-10-08
### Added
- 适配两码上传

## 2021-10-01
### Added
- 支持 GitHub Actions 填报
### Changed
- “抢排名”适配服务器凌晨维护

## 2021-09-10
### Fixed
- 变量名冲突导致的当发送消息失败时产生崩溃

## 2021-07-30
### Remove
- 移除填报每日两报的功能
### Changed
- “抢排名”功能可选开启

## 2021-06-24
### Added
- 增加“推送加”消息推送接口
### Fixed
- Server 酱推送消息接口更新

## 2021-04-17
### Added
- 适配登录接口对密码的 RSA 加密

## 2021-04-15
### Fixed
- `__EVENTTARGET` 参数更新
### Deprecated
- 健康之路不再需要每日两报，未来将移除相关适配代码

## 2021-02-28
### Fixed
- 修复新增的`是否在校`，`是否在家`字段

## 2021-02-03
### Fixed
- 修复登录次数限制

## 2021-01-29
### Added
- “抢排名”功能上线

## 2021-01-28
### Added
- 填报每日两报功能支持

## 2021-01-27
### Added
- 基础填报功能实现
- Server 酱消息推送
- 支持命令行添加用户


pcr.py
# -*- coding: UTF-8 -*-
import hashlib
import os
import logging
from utils import abs_path, getUsers, login, html2JsLine, jsLine2Json, getSendApi, sendMsg, getTime

config = abs_path + '/pcr.yaml'
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(message)s')


def get_last_pcr_result_md5():
    result_path = abs_path + '/pcr_result_md5.txt'
    if not os.path.exists(result_path):
        return None
    with open(result_path, 'r', encoding='utf-8') as f:
        last_pcr_result_md5 = f.read()
    return last_pcr_result_md5


def save_new_pcr_result_md5(pcr_result_md5):
    with open(abs_path + '/pcr_result_md5.txt', 'w', encoding='utf-8') as f:
        f.write(pcr_result_md5)


def check_PCR_result():
    users = getUsers(config)
    username = list(users.keys())[0]
    password = users[username][0]
    session = login(username, password)
    if not session:
        logging.info('登录失败')
        return False
    pcr_url = 'https://selfreport.shu.edu.cn/HSJC/JianKEDA.aspx'
    pcr_index = session.get(pcr_url).text
    pcr_line = html2JsLine(pcr_index)
    pcr_result = None
    for i, h in enumerate(pcr_line):
        if 'Panel1_P_HeSJCJL_HeSJCList' in h:
            pcr_result = jsLine2Json(pcr_line[i - 1])['F_Items']
            break
    if pcr_result is None:
        logging.info('获取PCR结果失败')
        return False
    last_pcr_result_md5 = get_last_pcr_result_md5()
    new_pcr_result_md5 = hashlib.md5(str(pcr_result).encode('utf-8')).hexdigest()
    if last_pcr_result_md5 == new_pcr_result_md5:
        logging.info('PCR结果未更新')
        return False
    updated_msg = 'PCR结果已更新'
    logging.info(updated_msg)
    send_api = getSendApi(config)
    now = getTime().strftime('%Y-%m-%d %H:%M:%S')
    send_result = sendMsg(updated_msg, now + '\n\n' + updated_msg, send_api['api'], send_api['key'])
    if send_result:
        logging.info('消息发送成功，更新核酸结果哈希值')
        save_new_pcr_result_md5(new_pcr_result_md5)
    else:
        logging.info('消息发送失败，未更新核酸结果哈希值')
    return True


if __name__ == '__main__':
    if check_PCR_result():
        exit(0)
    else:
        exit(1)


ovpn.py
import os
import re
SHU_DOMAINS = [
    'speedtest.shu.edu.cn',
    'selfreport.shu.edu.cn',
    'newsso.shu.edu.cn',
    'oauth.shu.edu.cn',
    'services.shu.edu.cn',
    'xk.autoisp.shu.edu.cn',
    'xk.shu.edu.cn',
]
SHU_IPS = []


def get_ip(domain):
    ip = os.popen("dig +short {}".format(domain)).read().strip()
    ip = re.findall(r'\d+\.\d+\.\d+\.\d+', ip)[0]
    return ip


def get_route_config():
    global SHU_IPS
    routing_config = "route-nopull\nroute-metric 150\nmax-routes 1000"
    for domain in SHU_DOMAINS:
        ip = get_ip(domain)
        SHU_IPS.append(ip)
    SHU_IPS = list(set(SHU_IPS))
    for ip in SHU_IPS:
        routing_config += "\n"
        routing_config += "route %s 255.255.255.255" % ip
    print(routing_config)
    return routing_config


def config_ovpn():
    global SHU_IPS
    users = os.environ['users'].split(';')
    user = users[0].split(',')
    secret = user[0] + "\n" + user[1]
    with open('secret.txt', 'w') as f:
        f.write(secret)

    routing_config = get_route_config()

    with open('.github/vpn/config.ovpn', 'r') as f:
        content = f.read()
        content = content.replace('# ROUTING CONFIG', routing_config)
        with open('.github/vpn/config.ovpn', 'w') as fa:
            fa.write(content)


if __name__ == '__main__':
    print('This Project is No Longer Maintained. ')
    exit(0)
    config_ovpn()


ovpn.sh
#!/usr/bin/env bash
connect_times=0
while true; do
  openvpn --config ".github/vpn/config.ovpn" --log "vpn.log" --daemon
  connect_times=$((connect_times + 1))
  sleep 5
  test_times=0
  while true; do
    if ping -c1 xk.autoisp.shu.edu.cn; then
      echo "OpenVPN已成功连接"
      exit 0
    else
      test_times=$((test_times + 1))
      if [ $test_times -gt 15 ]; then
        echo "测试OpenVPN连接失败15次，将尝试重新连接......"
        if killall openvpn; then
          echo "已停止OpenVPN进程"
        else
          echo "停止OpenVPN进程失败"
        fi
        break
      fi
      sleep 2
    fi
  done

  if [ $connect_times -gt 3 ]; then
    echo "OpenVPN连接失败3次，将退出"
    exit 1
  fi

  if [ $connect_times -gt 1 ]; then
    sleep_time=$((RANDOM % 240 + 60))
  else
    sleep_time=$((RANDOM % 900 + 300))
  fi
  echo "开始休眠 $sleep_time 秒"
  for ((i = 0; i < sleep_time; i = i + 10)); do
    echo -ne "休眠剩余：$((sleep_time - i))         \r"
    sleep 10
  done
done


ag.py
# -*- coding: UTF-8 -*-
import os
import re
import json
import math
import base64
import random
import logging
import datetime
from time import sleep
from PIL import Image, ImageOps, ImageEnhance
from utils import abs_path, getUsers, login, html2JsLine, jsLine2Json, getTime, sendMsg, getSendApi

AG_IMG_PATH = abs_path + '/ag_img'
if not os.path.exists(AG_IMG_PATH) or len(os.listdir(AG_IMG_PATH)) == 0:
    raise Exception('ag_img not exists')

if os.path.exists(abs_path + '/ag.yaml'):
    config = abs_path + '/ag.yaml'
else:
    config = abs_path + '/config.yaml'
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(message)s')


def _sleep():
    os_sleep = os.getenv('sleep_time', '')
    os_sleep = 5 if os_sleep == '' else os_sleep
    if os_sleep == 'random':
        sleep_time = random.randint(2 * 60, 30 * 60)
    else:
        sleep_time = int(os_sleep)
    logging.info(f'休眠 {sleep_time}s，约{sleep_time // 60}分钟')
    sleep(sleep_time)
    logging.info("休眠结束")


def get_random_ag_img(img_path):
    json_path = f'{img_path}/ag_img_times.json'
    if not os.path.exists(json_path):
        times = {}
    else:
        with open(json_path, 'r') as f:
            times = json.load(f)
    new_times = {}
    for img_name in os.listdir(img_path):
        if 'compress' not in img_name:
            if img_name not in times.keys():
                new_times[img_name] = 0
            else:
                new_times.update({img_name: times[img_name]})
    sorted_times = sorted(new_times.items(), key=lambda x: x[1], reverse=False)
    min_times_imgs = [x[0] for x in sorted_times if x[1] == sorted_times[0][1]]
    random_img = random.choice(min_times_imgs)
    new_times[random_img] += 1
    with open(json_path, 'w', encoding='utf-8') as f:
        json.dump(new_times, f)
    return f'{img_path}/{random_img}'


def compress_img(img_path):
    cur_img = Image.open(img_path)
    cur_img = ImageOps.exif_transpose(cur_img)
    raw_width, raw_height = cur_img.size
    angle = random.randint(1, 6)
    cur_img = cur_img.rotate(angle, expand=True)
    new_width, new_height = cur_img.size
    blank_width = int(math.tan(math.radians(angle)) * raw_height + 10)
    blank_height = int(math.tan(math.radians(angle)) * raw_width + 10)
    left = blank_width
    top = blank_height
    right = new_width - blank_width
    bottom = new_height - blank_height
    cur_img = cur_img.crop((left, top, right, bottom))

    width, height = cur_img.size
    crop_width = width * 0.96
    crop_height = height * 0.96
    left = random.randint(0, int(width - crop_width))
    top = random.randint(0, int(height - crop_height))
    right = left + crop_width
    bottom = top + crop_height
    cur_img = cur_img.crop((left, top, right, bottom))

    bri_enhancer = ImageEnhance.Brightness(cur_img)
    cur_img = bri_enhancer.enhance(random.randint(8, 12) / 10)
    col_enhancer = ImageEnhance.Color(cur_img)
    cur_img = col_enhancer.enhance(random.randint(9, 11) / 10)

    cps_time = getTime().strftime('%M%S%f')
    new_img_path = img_path.replace('.jpg', f'_{cps_time}_compress.jpg')
    target_size = 3 * 1024 * 1024
    quality = 90
    step = 5

    cur_img.save(new_img_path)
    cur_img_size = os.path.getsize(new_img_path)
    while cur_img_size > target_size and quality > 10:
        cur_img.save(new_img_path, quality=quality, optimize=True)
        cur_img = Image.open(new_img_path)
        cur_img_size = os.path.getsize(new_img_path)
        quality -= step

    return new_img_path


class AgUpload:
    def __init__(self, username, password):
        self.username = username
        self.password = password
        self.session = None
        self._login()
        self.id_num = username

        self.img_path = None
        self.archive_img_name = None
        self.img = None

        self.view_state = None
        self.view_state_generator = None
        self.t = getTime()
        self.t -= datetime.timedelta(minutes=2)
        self.test_times = None
        self.test_check = None
        self.report_form = None
        self.file_form = None

    def _login(self):
        self.session = login(self.username, self.password)
        if not self.session:
            logging.error(f'登录失败，用户名：{self.username}')
            return False
        return True

    def _archive_img(self):
        archive_path = f'{abs_path}/ag_archive'
        if not os.path.exists(archive_path):
            os.mkdir(archive_path)
        archive_id_path = f'{archive_path}/{self.id_num}'
        if not os.path.exists(archive_id_path):
            os.mkdir(archive_id_path)
        new_img_path = f'{archive_id_path}/{self.archive_img_name}'
        with open(self.img_path, 'rb') as f:
            img_data = f.read()
        with open(new_img_path, 'wb') as f:
            f.write(img_data)

    def _read_notice(self, notice_url):
        notice_html = self.session.get(url=notice_url).text
        notice_event_target = re.search(r'Submit\',name:\'(.*?)\',disabled:true', notice_html).group(1)
        notice_form = {
            '__EVENTTARGET': notice_event_target,
            '__EVENTARGUMENT': '',
            '__VIEWSTATE': self.view_state,
            '__VIEWSTATEGENERATOR': self.view_state_generator,
            'F_TARGET': 'p1_ctl01_btnSubmit',
            'p1_ctl00_Collapsed': 'false',
            'p1_Collapsed': 'false',
            'F_STATE': 'eyJwMV9jdGwwMCI6eyJJRnJhbWVBdHRyaWJ1dGVzIjp7fX0sInAxIjp7IklGcmFtZUF0dHJpYnV0ZXMiOnt9fX0=',
        }
        notice_result = self.session.post(url=notice_url, data=notice_form)
        if 'HeSJCSelfUploads.aspx' in notice_result.url:
            return True
        else:
            return False

    def _get_report_form(self):
        ag_url = 'https://selfreport.shu.edu.cn/HSJC/HeSJCSelfUploads.aspx'
        ag_html = self.session.get(url=ag_url).text

        report_even_target = 'p1$P_Upload$btnUploadImage'
        self.view_state = re.search(r'id="__VIEWSTATE" value="(.*?)" /', ag_html).group(1)
        self.view_state_generator = re.search(r'id="__VIEWSTATEGENERATOR" value="(.*?)" /', ag_html).group(1)
        t = getTime()
        t -= datetime.timedelta(minutes=2)
        self.t = t
        test_date = t.strftime('%Y-%m-%d %H:%M')
        self.test_times = "1" if getTime().hour < 12 else "2"
        self.test_check = f'当天第{self.test_times}次({t.year}/{t.month}/{t.day}'
        self.archive_img_name = f'IMG_{t.strftime("%Y%m%d_%H%M%S")}.jpg'

        name = ""
        ag_line = html2JsLine(ag_html)
        for i, line in enumerate(ag_line):
            if 'p1_XingMing' in line:
                name = jsLine2Json(ag_line[i - 1])['Text']
                break

        ag_b64str = 'eyJwMV9Hb25nSGFvIjogeyJUZXh0IjogIiJ9LCAicDFfWGluZ01pbmciOiB7IlRleHQiOiAiIn0sICJwMV9QX1VwbG9hZF9DaGVuZ051byI6IHsiQ2hlY2tlZCI6IHRydWV9LCAicDFfUF9VcGxvYWRfY3RsMDAiOiB7IklGcmFtZUF0dHJpYnV0ZXMiOiB7fX0sICJwMV9QX1VwbG9hZF9TaGVuVFpLIjogeyJGX0l0ZW1zIjogW1si5ZCmIiwgIjxzcGFuIHN0eWxlPSdjb2xvcjpncmVlbic+5peg5Lul5LiK55eH54q2KE5vKTwvc3Bhbj4iLCAxXSwgWyLmmK8iLCAiPHNwYW4gc3R5bGU9J2NvbG9yOnJlZCc+5pyJ5Lul5LiK55eH54q25LmL5LiAKFllcyk8L3NwYW4+IiwgMV1dLCAiU2VsZWN0ZWRWYWx1ZSI6ICLlkKYifSwgInAxX1BfVXBsb2FkX0ppYW5DTFgiOiB7IkZfSXRlbXMiOiBbWyLmipfljp8iLCAiPHNwYW4gc3R5bGU9J2ZvbnQtd2VpZ2h0OmJvbGRlcjsnPuaKl+WOnyhBbnRpZ2VuIFRlc3QpPC9zcGFuPiIsIDFdLCBbIuaguOmFuCIsICLmoLjphbgoTnVjbGVpYyBBY2lkIFRlc3QpIiwgMV1dLCAiU2VsZWN0ZWRWYWx1ZSI6ICLmipfljp8ifSwgInAxX1BfVXBsb2FkX0NhaVlGUyI6IHsiRl9JdGVtcyI6IFtbIum8u+iFlOaLreWtkCIsICLpvLvohZTmi63lrZAoTm9zZSkiLCAxXSwgWyLpvLvlkr3mi63lrZAiLCAi6by75ZK95out5a2QKE5vc2UrVGhyb2F0KSIsIDFdLCBbIuWPo+iFlOaLreWtkCIsICLlj6PohZTmi63lrZAoVGhyb2F0KSIsIDFdXSwgIlNlbGVjdGVkVmFsdWUiOiAi6by76IWU5out5a2QIn0sICJwMV9QX1VwbG9hZF9IZVNKQ1JRIjogeyJUZXh0IjogIiJ9LCAicDFfUF9VcGxvYWRfQ2lTaHUiOiB7IkZfSXRlbXMiOiBbWyIxIiwgIuesrDHmrKEoRmlyc3QpIiwgMV0sIFsiMiIsICLnrKwy5qyhKFNlY29uZCkiLCAxXSwgWyIzIiwgIuesrDPmrKEoVGhpcmQpIiwgMV1dLCAiU2VsZWN0ZWRWYWx1ZSI6ICIxIn0sICJwMV9QX1VwbG9hZF9KaWFuQ0pHIjogeyJGX0l0ZW1zIjogW1si6Zi05oCnIiwgIjxzcGFuIHN0eWxlPSdjb2xvcjpncmVlbic+6Zi05oCnKE5lZ2F0aXZlKTwvc3Bhbj4iLCAxXSwgWyLpmLPmgKciLCAiPHNwYW4gc3R5bGU9J2NvbG9yOnJlZCc+6Ziz5oCnKFBvc2l0aXZlKTwvc3Bhbj4iLCAxXSwgWyLml6DmlYgiLCAi5peg5pWIKEludmFsaWQpIiwgMV0sIFsi5pqC5peg57uT5p6cIiwgIuaaguaXoOe7k+aenChObyBSZXN1bHQpIiwgMV1dLCAiU2VsZWN0ZWRWYWx1ZSI6ICLpmLTmgKcifSwgInAxX1BfVXBsb2FkIjogeyJJRnJhbWVBdHRyaWJ1dGVzIjoge319LCAicDFfR3JpZERhdGEiOiB7IlJlY29yZENvdW50IjogMCwgIkZfUm93cyI6IFtdLCAiSUZyYW1lQXR0cmlidXRlcyI6IHt9fSwgInAxIjogeyJJRnJhbWVBdHRyaWJ1dGVzIjoge319LCAiV19TaG93UGljIjogeyJJRnJhbWVBdHRyaWJ1dGVzIjoge319fQ=='
        ag_json = json.loads(base64.b64decode(ag_b64str).decode('utf-8'))
        ag_json['p1_GongHao']['Text'] = self.id_num
        ag_json['p1_XingMing']['Text'] = name
        ag_json['p1_P_Upload_HeSJCRQ']['Text'] = test_date
        ag_json['p1_P_Upload_CiShu']['SelectedValue'] = self.test_times
        fstate = base64.b64encode(json.dumps(ag_json, ensure_ascii=False).encode("utf-8")).decode("utf-8")

        self.report_form = {
            '__EVENTTARGET': report_even_target,
            '__EVENTARGUMENT': '',
            '__VIEWSTATE': self.view_state,
            '__VIEWSTATEGENERATOR': self.view_state_generator,
            'p1$P_Upload$ChengNuo': 'p1_P_Upload_ChengNuo',
            'p1$P_Upload$ShenTZK': b'\xe5\x90\xa6'.decode(),
            'p1$P_Upload$JianCLX': b'\xe6\x8a\x97\xe5\x8e\x9f'.decode(),
            'p1$P_Upload$CaiYFS': b'\xe9\xbc\xbb\xe8\x85\x94\xe6\x8b\xad\xe5\xad\x90'.decode(),
            'p1$P_Upload$HeSJCRQ': test_date,
            'p1$P_Upload$CiShu': self.test_times,
            'p1$P_Upload$JianCJG': b'\xe9\x98\xb4\xe6\x80\xa7'.decode(),
            'p1_P_Upload_ctl00_Collapsed': 'false',
            'p1_P_Upload_Collapsed': 'false',
            'p1_GridData_Collapsed': 'false',
            'p1_GridData_HiddenColumns': '["p1_GridData_ctl00"]',
            'p1_Collapsed': 'false',
            'W_ShowPic_Collapsed': 'false',
            'W_ShowPic_Hidden': 'true',
            'F_STATE': fstate,
            'F_TARGET': 'p1_P_Upload_btnUploadImage',
            'X-FineUI-Ajax': 'true',
        }

    def _get_img_file(self):
        img_path = get_random_ag_img(AG_IMG_PATH)
        img_path = compress_img(img_path)
        self.img = open(img_path, 'rb')
        self.file_form = {
            'p1$P_Upload$FileHeSJCBG': (self.archive_img_name, self.img, 'image/jpeg', {'Content-Type': 'image/jpeg'}),
        }
        self.img_path = img_path

    def upload(self):
        now = self.t.strftime('%Y-%m-%d %H:%M:%S')
        if self.session is None:
            title = f'{self.id_num}登录失败'
            logging.error(title)
            return 'fail', f'{now}\n\n{title}'

        self._get_report_form()
        self._get_img_file()
        title = f'{self.id_num}的第{self.test_times}次结果'
        ag_upload = 'https://selfreport.shu.edu.cn/HSJC/HeSJCSelfUploads.aspx'
        ag_html = self.session.get(ag_upload).text
        if self.test_check in ag_html:
            title += '已上传过'
            logging.info(title)
            self.img.close()
            os.remove(self.img_path)
            return 'uploaded', f'{now}\n\n{title}'

        upload_times = 0
        result_list = []
        while True:
            sleep(5)
            result = self.session.post(url=ag_upload, data=self.report_form, files=self.file_form).text
            upload_times += 1
            if '上传成功' in result or self.test_check in result or '更新失败' in result or upload_times >= 3:
                break
            elif '.aspx' in result.split('&#39;')[1]:
                notice_url = 'https://selfreport.shu.edu.cn' + result.split('&#39;')[1]
                self._read_notice(notice_url)
            logging.info(result)
            result_split = result.split('F.alert')
            if len(result_split) > 1:
                result_split = result_split[-1].split('&#39;')[1]
                result_list.append(result_split)
            self._get_report_form()
            self.img.close()
            os.remove(self.img_path)
            self._get_img_file()

        self.img.close()

        if '上传成功' in result or self.test_check in result:
            title += '上传成功'
            return_ = 'success', f'{now}\n\n{title}'
            self._archive_img()
        elif '更新失败' in result:
            title += '已上传过'
            return_ = 'uploaded', f'{now}\n\n{title}'
        else:
            title += '上传失败'
            logging.info(result)
            result = result.split('F.alert')
            if len(result) > 1:
                result = result[-1].split('&#39;')[1]
                result_list.append(result)
            result_str = '\n\n'.join(result_list)
            return_ = 'fail', f'{now}\n\n{title}\n\n{result_str}'
        logging.info(title)
        os.remove(self.img_path)
        return return_


class Main:
    def __init__(self):
        self.SUCCESS = []
        self.UPLOADED = []
        self.FAIL = []
        self.status_map = {
            'success': self.SUCCESS,
            'uploaded': self.UPLOADED,
            'fail': self.FAIL,
        }

    def run(self):
        users = getUsers(config)
        users = [(username, password) for username in users for password in users[username]]
        random.shuffle(users)
        for i, j in enumerate(users):
            username, password = j
            status, desp = AgUpload(username, password).upload()
            self.status_map[status].append(desp)
            if i < len(users) - 1:
                sleep_time = random.randint(5, 10)
                logging.info(f'休眠{sleep_time}分钟')
                sleep(sleep_time * 60)

        send_api = getSendApi(config)
        title = 'AgUpload'
        title += f'{len(self.SUCCESS)}个成功' if self.SUCCESS else ''
        title += f'{len(self.UPLOADED)}个已上传过' if self.UPLOADED else ''
        title += f'{len(self.FAIL)}个失败' if self.FAIL else ''
        desp = '\n\n'.join(self.SUCCESS + self.UPLOADED + self.FAIL)
        send_result = sendMsg(title, desp, send_api['api'], send_api['key'])
        logging.info('消息发送成功') if send_result else logging.info('消息发送失败')
        if len(self.FAIL):
            return False
        return True


if __name__ == '__main__':
    _sleep()
    if Main().run():
        exit(0)
    exit(1)
